### ISSUE TITLE
Liquid {% include 'another-file' %} and FileNotFoundException

### ISSUE THREAD

**Issue by jagbarcelo** [OrchardCMS/OrchardCore#7353]
Trying to find an alternate solution to my problems with #7352, I have found this [hex2rgb liquid converter](https://github.com/granteagon/Shopify-hex2rgb.liquid)

I have copied the file hex2rgb.liquid file side by side with the rest of my liquid templates in **myTheme/Views**. However, whenever I try to include it using:
`{% include "hex2rgb" %}`
the page where it should be used does not work anymore and the following exception occurs:

![image](https://user-images.githubusercontent.com/6704109/96866633-ddca2200-146b-11eb-8baa-d0b3883aabdc.png)

Should I place the file to be included somewhere in particular? Where is OC looking for it that it does not find it?

I have already tried several combinations, including (and not including) the extension and the directory where the file resides.

Raw exception details follow:

```
System.IO.FileNotFoundException: hex2rgb.liquid
   at Fluid.Ast.IncludeStatement.WriteToAsync(TextWriter writer, TextEncoder encoder, TemplateContext context)
   at Fluid.Ast.IfStatement.WriteToAsync(TextWriter writer, TextEncoder encoder, TemplateContext context)
   at Fluid.BaseFluidTemplate`1.Awaited(ValueTask`1 task, TextWriter writer, TextEncoder encoder, TemplateContext context, List`1 statements, Int32 startIndex)
   at OrchardCore.DisplayManagement.Liquid.LiquidViewTemplate.RenderAsync(RazorPage`1 page) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement.Liquid\LiquidViewTemplate.cs:line 134
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 297
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 174
   at AspNetCore.Views_FlowPart.ExecuteAsync() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Flows\Views\FlowPart.cshtml:line 23
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 297
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 174
   at CallSite.Target(Closure , CallSite , Object )
   at OrchardCore.DisplayManagement.Zones.ZoneShapes.ColumnGrouping(Object DisplayAsync, Object Shape, IShapeFactory ShapeFactory) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Zones\ZoneShapes.cs:line 305
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 297
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 174
   at CallSite.Target(Closure , CallSite , Object )
   at OrchardCore.DisplayManagement.Zones.ZoneShapes.CardGrouping(Object DisplayAsync, Object Shape, IShapeFactory ShapeFactory) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Zones\ZoneShapes.cs:line 204
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 297
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 174
   at CallSite.Target(Closure , CallSite , Object )
   at OrchardCore.DisplayManagement.Zones.ZoneShapes.ContentZone(Object DisplayAsync, Object Shape, IShapeFactory ShapeFactory) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Zones\ZoneShapes.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 297
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 174
   at OrchardCore.DisplayManagement.Liquid.Filters.LiquidViewFilters.<ShapeRender>g__Awaited|11_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement.Liquid\Filters\LiquidViewFilters.cs:line 114
   at Fluid.Ast.FilterExpression.EvaluateAsync(TemplateContext context)
   at Fluid.Ast.OutputStatement.<WriteToAsync>g__Awaited|7_0(ValueTask`1 t, TextWriter w, TextEncoder enc, TemplateContext ctx)
   at Fluid.BaseFluidTemplate`1.Awaited(ValueTask`1 task, TextWriter writer, TextEncoder encoder, TemplateContext context, List`1 statements, Int32 startIndex)
   at OrchardCore.DisplayManagement.Liquid.LiquidViewTemplate.RenderAsync(RazorPage`1 page) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement.Liquid\LiquidViewTemplate.cs:line 134
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.<ProcessAsync>g__Awaited|12_0(Task`1 task) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 297
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 116
   at OrchardCore.DisplayManagement.Implementation.DefaultHtmlDisplay.ExecuteAsync(DisplayContext context) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultHtmlDisplay.cs:line 174
   at CallSite.Target(Closure , CallSite , Object )
   at AspNetCore.Views_Item_Display.ExecuteAsync() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Contents\Views\Item\Display.cshtml:line 7
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeResultFilters>g__Awaited|27_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapWhenMiddleware.Invoke(HttpContext context)
   at OrchardCore.Apis.GraphQL.GraphQLMiddleware.Invoke(HttpContext context, IAuthorizationService authorizationService, IAuthenticationService authenticationService, ISchemaFactory schemaService) in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Apis.GraphQL\GraphQLMiddleware.cs:line 47
   at SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Diagnostics.DiagnosticsStartupFilter.<>c__DisplayClass3_0.<<Configure>b__1>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.Diagnostics\DiagnosticsStartupFilter.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at OrchardCore.ContentPreview.PreviewStartupFilter.<>c.<<Configure>b__1_1>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore.Modules\OrchardCore.ContentPreview\PreviewStartupFilter.cs:line 23
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 63
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 236
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 59
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by Skrypt** [comment]
This should definitely work. Try the /Views/Share folder and/or make sure your file name doesn't have a typo..

**Comment by jagbarcelo** [comment]
No, there is no typo. I have also tried moving the file into the /Views/Shared directory with the same results.

When we use `{% include ...` we get **FileNotFoundException**: With or without extension: same results. The .liquid extension is automatically added if none is provided.

If we try with your alternative `{{ "hex2rgb" | shape_new | shape_render }}` the exception we receive is **Exception: Shape type 'hex2rgb' not found**

![image](https://user-images.githubusercontent.com/6704109/96999961-58f80a80-1536-11eb-96a1-9ff376e4087a.png)

**Comment by Skrypt** [comment]
Wait what? You have something wrong in your dependencies because here it should create a Shape on the fly right in the View.

**Comment by jtkech** [comment]
@Skrypt @jagbarcelo yes this is because of the `{% include %}` statement

The Fluid library uses the `TemplateContext.FileProvider` property that we don't initialize. We would need to initialize it with our custom file provider, so that the included file can be found in the project location in dev mode, or embedded in the module / theme assembly in prod mode.

Just tried, it works but you have to pass the full relative path (the virtual one, and case sensitive), e.g.

    {% include "Areas/TheBlogTheme/Views/Test" %}

I will do a PR soon

**Comment by Skrypt** [comment]
So why is it working for my in a custom theme then?

**Comment by jptissot** [comment]
I don't think you are using include. You are rendering a shape (I think)

**Comment by Skrypt** [comment]
Yeah but @jagbarcelo said it is not working for him. Just like if the` | shape_new` was not working for him.

**Comment by jtkech** [comment]
Yes, i just tried and then fixed the original issue related to the include statement

@Skrypt but yes you're right `{{ "hex2rgb" | shape_new | shape_render }}` should work, it may fail depending on what the file contains, but you're right the file should be found ;) need to leave but i will confirm it tomorrow night.

**Comment by jagbarcelo** [comment]
In our custom theme 'myTheme' we have several liquid files under `/Views/` such as `Layout.liquid` or `Content.liquid`.
The `include` and ` | shape_new` I was trying to use was inside a liquid file called `Widget-Carousel.liquid` which, as you guess, is a Widget.
The only particular thing about this Widget is that it was not created in `myTheme` from a recipe, but directly in the live website. I do not know if that would make any difference. I thought that, as long as a Widget with the name Carousel exists, the `Widget-Carousel.liquid` file should be applied, no matter if it comes from a theme or elsewhere.

@jtech: Thanks for fixing this. I have not yet applied your PR in our solution so I cannot yet confirm if it works in our case. However, I have tried your workaround using `{% include "Areas/myTheme/Views/hex2rgb" %}` and it does not work either. We do not have any **Areas** directory nor in myTheme project nor in the website project where myTheme is applied.

Here is a capture of out top-level directories in myTheme project:
![image](https://user-images.githubusercontent.com/6704109/97070804-8945b480-15db-11eb-9bfe-e550d055f291.png)

And here are the two files involved in the issue (both selected):
![image](https://user-images.githubusercontent.com/6704109/97070884-1a1c9000-15dc-11eb-8132-d304da054c6d.png)

Are we missing something in the tree structure? Thanks all of you for your help

**Comment by jtkech** [comment]
@jagbarcelo yes, for the include statement fix that was merged you need to fork the last dev branch, or reference the packages from the cloudsmith feed related to the dev branch, Nuget feed is related to the master branch.

> We do not have any Areas directory

There are "virtual" folders that we always have and use e.g. `Areas/MyTheme/Views/...`, there are related to embedded files in modules / themes assemblies. In dev mode we map this virtual path to the physical location of the file, in prod we map it to the embedded file of the right assembly, but in both case we first use the virtual path. Note: you don't see it for static assets as we have static file providers where you don't have to specify the `Areas` and `wwwroot` segments.

About the shape liquid filters, just tried in our `TheBlogTheme` a simple `Hello world` in a `hex2rgb.liquid` file, and the following in `Widget-RawHtml.liquid`

    {{ "hex2rgb" | shape_new | shape_render }}
    {{ Model.ContentItem.Content.RawHtml.Content.Html | shortcode | raw }}

As said by @Skrypt it works, it's like the `hex2rgb.liquid` shape was not harvested, try to do a rebuild on your theme, also check in the solution if this file is marked as embedded resource, it is done automatically but may be not if copy pasted from another place. This is necessary because, even if in dev mode we use physical files, as said we first use virtual paths that are only enlisted for embedded resources.

Otherwise we would need more info on your context, first are you using the full source code from the dev branch, or referencing packages from Nuget / cloudsmith?

**Comment by jagbarcelo** [comment]
We are referencing the dev branch (cloudsmith.io), but we have it updated only up to 1.0.0-rc2-14880.

In our website .csproj file we have:
`<PackageReference Include="OrchardCore.Application.Cms.Targets" Version="1.0.0-rc2-14880" />`
By the way, myTheme.csproj is also referenced in the website .csproj:
`<ProjectReference Include="..\myTheme\myTheme.csproj" />`

One of the thinks that I first checked was if `hex2rgb.liquid` was marked as an *Embedded resource* in our Theme (and it is).
![image](https://user-images.githubusercontent.com/6704109/97091172-44049f80-163a-11eb-8f98-cac2ff5aaaa8.png)

We can rule out the shape problem. I have now tried again, adding
```
{% assign hex2rgb_hex = bgcolor %}
{{ "hex2rgb" | shape_new | shape_render }}
```
to our `Widget-Carousel.liquid` file and the shape is called and run. I know because I have put a *Hello world* there too (red circles):
![image](https://user-images.githubusercontent.com/6704109/97091444-2df7de80-163c-11eb-968b-7425012f4fc6.png)
and the text is included in the output (follow the red arrow).

However, the script is not executed as expected (see the green circle). It seems `{{ include ... }}` does not get executed the same way as `{{ ... | shape_new | shape_render }}` is.

Right now, the only workaround we have to run it correctly is to include the whole file manually right in the place we want it (instead of using the `{{ include ... }}` command). The right output should be this one (in green too).
![image](https://user-images.githubusercontent.com/6704109/97091642-93000400-163d-11eb-9ac3-b46e88327a11.png)

Can you find any reason why the `{ ... | shape_new | shape_render }` command gets executed, but only partially? It does not properly executes the `{% assign ... }` and the rest of commands. Maybe they get executed, but the variables are local and not available at the caller to be used later, as opposed to an `{{ include ... }}` ?

**Comment by jtkech** [comment]
@jagbarcelo 

Okay good to know that the shape is executed

>  Maybe they get executed, but the variables are local and not available at the caller to be used later, as opposed to an {{ include ... }} ?

You're right, that's the reason

So in your case you need the include statement, when you will able to use the last dev, let me know if it works for you

**Comment by jagbarcelo** [comment]
@jtkech Good news

We have just updated our project to OC 1.0.0-rc2-15201 (today's dev version) and I can confirm that your PR works, with a little tweak that I'll explain later (along with another question).

 We have first tried to use the straigth simple version:
```
{% assign hex2rgb_hex = bgcolor %}
{% include "hex2rgb" %}
```
which still throws an exception.

The second time we used:
```
{% assign hex2rgb_hex = bgcolor %}
{% include "Areas/myTheme/Views/hex2rgb" %}
```
and the file to be included was found and executed, but the output variables `hex2rgb_r`, `hex2rgb_g` and `hex2rgb_b` (which were declared **inside** the include) were empty when they were read later on, and thus the output html still contained:

![image](https://user-images.githubusercontent.com/6704109/97679480-99570b80-1a95-11eb-80a1-dc4025ef93be.png)

The third time we used:
```
{% assign hex2rgb_hex = bgcolor %}
{% assign hex2rgb_r = 0 %}
{% assign hex2rgb_g = 0 %}
{% assign hex2rgb_b = 0 %}
{% include "Areas/myTheme/Views/hex2rgb" %}
```
to initialise the variables **outside and before** the include and this time, the output was right:

![image](https://user-images.githubusercontent.com/6704109/97679782-fe126600-1a95-11eb-8c33-3be3a3f0550c.png)

So my question now is this: Should the include statement behave **exactly** as if the actual contents of the included file were inserted in that exact place of the file? If that's the case, then I shouldn't need to declare those three variables before and outside the include statement.
Or... is this the expected behaviour? Variables declared in an included file are kept local to that file (and not avaiblable to the original file making the include).
Depending on your answer to this, there might be another minor issue around to be solved. Regards.
